extensions/firmware/org.eclipse.smarthome.firmware.filesystem/src/main/java/org/eclipse/smarthome/firmware/filesystem/FileSystemFirmwareProvider.java

```diff
@@ -0,0 +1,291 @@
+/**
+ * Copyright (c) 2016 Deutsche Telekom AG and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.firmware.filesystem;
+
+import static java.nio.file.StandardWatchEventKinds.*;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchEvent.Kind;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Properties;
+import java.util.ResourceBundle;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.commons.io.FileUtils;
+import org.eclipse.smarthome.config.core.ConfigConstants;
+import org.eclipse.smarthome.core.service.AbstractWatchQueueReader;
+import org.eclipse.smarthome.core.service.AbstractWatchService;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.firmware.Firmware;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareProvider;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * <p>
+ * A simple {@link FirmwareProvider}, that retrieves {@link Firmware} from {@code userdata/firmware}.
+ * Each firmware consists of two files which must be named using the pattern shown below:
+ * A properties file containing the firmware metadata and an image file with the firmware image.
+ * Files are in sub folders according their bindingId/{@link ThingType}.
+ * </p>
+ * <p>
+ * Each file name must follow the rules for a Java class name. For a specific firmware, all files must
+ * have the same name. At least the following files are required (called FW1 in this example):
+ * </p>
+ * <ul>
+ * <li>An image file containing the firmware image; the name of the image file is contained in the properties file</li>
+ * <li>{@code FW1.properties} - containing the complete firmware meta data with locale specific properties in English
+ * </li>
+ * </ul>
+ * <p>
+ * The property file can contain the following values (version and image are mandatory):
+ * 
+ * <pre>
+ * version = V1.2-b18
+ * image = imageV12b18.bin
+ * model = Model A
```

Comment:
I am missing here something like a hash value about the binary file (here: image).
Maybe add here a image.hash.md5, or image.hash.sha1 property to indicate an expected value for a MD5 or SHA1 based hash value. This value can be optional, if not specified dont check, but if specified the image used can be checked if hash value matches